libursa/QueryGraph.cpp

```diff
@@ -29,40 +32,116 @@ QueryGraph QueryGraph::dual() const {
     return result;
 }
 
+// Helper class for InorderGraphVisitor. Remembers how many incoming edges
+// are there for every node, and how many were already processed.
+// Type T is a type of user-supplied state, and TM is its constructor.
+template <typename T, T TM()>
 class NodeState {
     std::vector<NodeId> ready_predecessors_;
     uint32_t total_predecessors_;
-    QueryResult state_;
+    T state_;
 
    public:
-    NodeState() : state_(QueryResult::everything()), total_predecessors_(0) {}
+    // Constructs a new instance of NodeState with provided constructor.
+    NodeState() : state_(TM()), total_predecessors_(0) {}
```

Comment:
Do we want a copy constructor? :thinking:

Reply:
Also, uniform initialization { }

Reply:
In our case, QueryResult is not copy constructible anyway. But fair point.